### PR TITLE
Issues/#3 cleanup swift additions

### PR DESCRIPTION
**Pull request by DevTchernov**
Data Source classes wrapped on Kotlin side, Swift code cleared

### CODE REVIEW

File: sample/ios-app/TestProj.xcodeproj/xcshareddata/xcschemes/TestProj.xcscheme [hunk 1]
```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Alex009) [inline sample/ios-app/TestProj.xcodeproj/xcshareddata/xcschemes/TestProj.xcscheme]: why scheme set as shared? it's necessary?

File: sample/ios-app/src/TestCollectionViewController.swift [hunk 2]
```diff
@@ -9,20 +9,13 @@ import MultiPlatformLibraryUnits
...
-        let dataSource = FlatUnitCollectionViewDataSource()
-        dataSource.setup(for: collectionView)
-
-        let units = testing.getUnits() as! [UIAnyCellUnitProtocol]
-        dataSource.units = units
-
-        self.dataSource = dataSource
+        dataSource = UnitsDataSource.Factory().create(forCollectionView: collectionView)
+        dataSource.unitItems = TestingCollection(unitFactory: self).getUnits()
+        collectionView.reloadData()
```
Comment (Alex009) [inline sample/ios-app/src/TestCollectionViewController.swift]: it's required now?

File: sample/ios-app/src/TestViewController.swift [hunk 3]
```diff
@@ -9,21 +9,14 @@ import MultiPlatformLibraryUnits
...
-        dataSource.setup(for: tableView)
-        
-        let units = testing.getUnits() as! [UITableViewCellUnitProtocol]
-        dataSource.units = units
-        
         tableView.tableFooterView = UIView()
-        self.dataSource = dataSource
+        dataSource = UnitsDataSource.Factory().create(forTableView: tableView)
+        dataSource.unitItems = Testing(unitFactory: self).getUnits()
+        tableView.reloadData()
```
Comment (Alex009) [inline sample/ios-app/src/TestViewController.swift]: it's required now?

File: sample/ios-app/src/TestViewController.swift [hunk 4]
```diff
@@ -36,14 +29,13 @@ extension TestViewController: TestingUnitFactory {
     func createSimpleUnit(id: Int64, title: String, itemData: ItemData?) -> TableUnitItem {
         // without R.swift
         return UITableViewCellUnit<SimpleCell>(
-            data: SimpleCell.CellModel(id: id, title: title),
-            reuseId: SimpleCell.reusableIdentifier(),
-            nibName: SimpleCell.xibName(),
-            configurator: nil
-        )
+            data: SimpleCell.CellModel(id: id, title: itemData?.data ?? ""),
```
Comment (Alex009) [inline sample/ios-app/src/TestViewController.swift]: why title now from `itemData`?

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt [hunk 5]
```diff
@@ -7,24 +7,25 @@ package dev.icerock.moko.units
 import kotlinx.cinterop.ExportObjCClass
 import platform.Foundation.NSIndexPath
 import platform.UIKit.UICollectionView
-import platform.UIKit.UICollectionViewCell
 import platform.UIKit.UICollectionViewDataSourceProtocol
 import platform.UIKit.row
 import platform.darwin.NSInteger
 import platform.darwin.NSObject
+import platform.UIKit.UICollectionViewCell as UICollectionViewCell
```
Comment (Alex009) [inline units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt]: it's necessary alias?

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt [hunk 6]
```diff
@@ -7,24 +7,25 @@ package dev.icerock.moko.units
...
 
 @ExportObjCClass
-class UnitCollectionViewDataSource(
-    private val collectionView: UICollectionView
-) : NSObject(), UICollectionViewDataSourceProtocol {
+class UnitCollectionViewDataSource internal constructor(collectionView: UICollectionView): NSObject(), UICollectionViewDataSourceProtocol {
```
Comment (Alex009) [inline units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt]: reformat please to:
```
class UnitCollectionViewDataSource internal constructor(
    collectionView: UICollectionView
): NSObject(), UICollectionViewDataSourceProtocol {
```
it's more readable

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt [hunk 7]
```diff
@@ -7,24 +7,25 @@ package dev.icerock.moko.units
...
     private val unitsRegistry = UnitsRegistry<UICollectionView, CollectionUnitItem>(collectionView)
     var unitItems: List<CollectionUnitItem>? = null
         set(value) {
             field = value
             if (value != null) unitsRegistry.registerIfNeeded(value)
-            collectionView.reloadData()
```
Comment (Alex009) [inline units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt]: i think it's should be automatic. later we just replace this `reloadData` to animated items update with diff by `itemId`

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitTableViewDataSource.kt [hunk 8]
```diff
@@ -14,17 +14,18 @@ import platform.darwin.NSInteger
 import platform.darwin.NSObject
 
 @ExportObjCClass
-class UnitTableViewDataSource(
-    private val tableView: UITableView
-) : NSObject(), UITableViewDataSourceProtocol {
+class UnitTableViewDataSource internal constructor(tableView: UITableView): NSObject(), UITableViewDataSourceProtocol {
```
Comment (Alex009) [inline units/src/iosMain/kotlin/dev/icerock/moko/units/UnitTableViewDataSource.kt]: reformat to 
```
class UnitTableViewDataSource internal constructor(
    tableView: UITableView
): NSObject(), UITableViewDataSourceProtocol {
```
please

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitsSource.kt [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+                    get() {
+                        return unitsSource.unitItems
+                    }
+                    set(value) {
+                        unitsSource.unitItems = value
+                    }
+            }
+        }
+    }
+}
```
Comment (Alex009) [inline units/src/iosMain/kotlin/dev/icerock/moko/units/UnitsSource.kt]: for more simple usage from swift side it's may be reworked to:
```kotlin
fun TableUnitsSource(forTableView: UITableView): TableUnitsSource {
  return object: TableUnitsSource {
    private val unitsSource = UnitTableViewDataSource(forTableView)
    
    override var unitItems: List<TableUnitItem>?
      get() {
        return unitsSource.unitItems
      }
      set(value) {
        unitsSource.unitItems = value
      }
    }
}
```
just as global scope function, without class.
and simillar for Collection

File: units/src/iosMain/swift/Core/Units/UICellUnit.swift [hunk 10]
```diff
@@ -3,45 +3,51 @@
  */
 
 import UIKit
+import MultiPlatformLibrary
```
Comment (Alex009) [inline units/src/iosMain/swift/Core/Units/UICellUnit.swift]: i think it's not needed anymore?

File: units/src/iosMain/swift/Core/Units/UICollectionViewCellUnit.swift [hunk 11]
```diff
@@ -0,0 +1,31 @@
+//
+//  UICollectionViewCellUnit.swift
+//  MultiPlatformLibraryUnits
+//
+//  Created by Andrey Tchernov on 16/01/2020.
+//
+
```
Comment (Alex009) [inline units/src/iosMain/swift/Core/Units/UICollectionViewCellUnit.swift]: replace copyright to 
```
/*
 * Copyright 2019 IceRock MAG Inc. Use of this source code is governed by the Apache 2.0 license.
 */
```

File: units/src/iosMain/swift/Core/Units/UICollectionViewCellUnit.swift [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+  public func register(intoView: Any?) {
+    guard let collectionView = intoView as? UICollectionView else { return }
+    collectionView.register(
+      UINib(nibName: self.nibName, bundle: self.bundle),
+      forCellWithReuseIdentifier: self.reusableIdentifier)
+  }
+  
+  public func bind(cell: UICollectionViewCell) {
+    guard let cell = cell as? Cell else { return }
+    configurator?(cell)
```
Comment (Alex009) [inline units/src/iosMain/swift/Core/Units/UICollectionViewCellUnit.swift]: configurator should be called on each `collectionCell forIndexPath` call?

File: units/src/iosMain/swift/Core/Units/UITableViewCellUnit.swift [hunk 13]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2019 IceRock MAG Inc. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+import Foundation
+import UIKit
+import MultiPlatformLibrary
+
+import Foundation
+import MultiPlatformLibrary
```
Comment (Alex009) [inline units/src/iosMain/swift/Core/Units/UITableViewCellUnit.swift]: duplication

File: units/src/iosMain/swift/Core/Units/UITableViewCellUnit.swift [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    
+    tableView.register(
+      UINib(nibName: self.nibName, bundle: self.bundle),
+      forCellReuseIdentifier: self.reusableIdentifier
+    )
+  }
+  
+  public func bind(cell_ cell: UITableViewCell) {
+    guard let cell = cell as? Cell else { return }
+    configurator?(cell)
```
Comment (Alex009) [inline units/src/iosMain/swift/Core/Units/UITableViewCellUnit.swift]: configurator should be called on each `tableCell forIndexPath` call?

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitsSource.kt [hunk 15]
```diff
@@ -0,0 +1,48 @@
+package dev.icerock.moko.units
```
Comment (Alex009) [inline units/src/iosMain/kotlin/dev/icerock/moko/units/UnitsSource.kt]: copyright header required here

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitsSource.kt [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+import platform.UIKit.UICollectionView
+import platform.UIKit.UITableView
+import platform.UIKit.UICollectionViewDataSourceProtocol
+
+interface TableUnitsSource {
+    var unitItems: List<TableUnitItem>?
+    val tableView: UITableView?
+}
+
+interface CollectionUnitsSource  {
```
Comment (Alex009) [inline units/src/iosMain/kotlin/dev/icerock/moko/units/UnitsSource.kt]: i think sources must be separated to:
`TableUnitsSource.kt` and `CollectionUnitsSource.kt` with interface and `create` function

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt [hunk 17]
```diff
@@ -14,17 +14,20 @@ import platform.darwin.NSInteger
 import platform.darwin.NSObject
 
 @ExportObjCClass
-class UnitCollectionViewDataSource(
-    private val collectionView: UICollectionView
```
Comment (Alex009) [inline units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt]: i suggest backward compatible change:
```kotlin
class UnitCollectionViewDataSource(
    private val collectionView: UICollectionView,
    private val reloadDataHandler: (UICollectionView, oldData: List<CollectionUnitItem>?, newData: List<CollectionUnitItem>?) -> Unit = { collectionView, _, _ -> collectionView.reloadData() }
)
```
and in setter of `unitItems`:
```kotlin
val old = field
field = value
if (value != null) unitsRegistry.registerIfNeeded(value)
reloadDataHandler(collectionView, old, value)
```

Comment (Alex009) [reply units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt]: and same to TableViewDataSource

Comment (Alex009) [reply units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt]: it's allow configure any reload handler from outside and have default implementation, same as in old version

File: units/src/iosMain/swift/Core/Units/UITableViewCellUnit.swift (hunk 14 above)
Comment (DevTchernov) [reply units/src/iosMain/swift/Core/Units/UITableViewCellUnit.swift]: Yes

File: units/src/iosMain/swift/Core/Units/UICollectionViewCellUnit.swift (hunk 12 above)
Comment (DevTchernov) [reply units/src/iosMain/swift/Core/Units/UICollectionViewCellUnit.swift]: Yes

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt [hunk 18]
```diff
@@ -13,18 +13,27 @@ import platform.UIKit.row
...
 
+typealias UICollectionViewReloadHandler = (UICollectionView, oldData: List<CollectionUnitItem>?, newData: List<CollectionUnitItem>?) -> Unit
+
 @ExportObjCClass
-class UnitCollectionViewDataSource(
-    private val collectionView: UICollectionView
-) : NSObject(), UICollectionViewDataSourceProtocol {
+class UnitCollectionViewDataSource internal constructor(
+    private val collectionView: UICollectionView,
+    private val reloadDataHandler: UICollectionViewReloadHandler
```
Comment (Alex009) [inline units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt]: here need default value, to simple usage from kotlin side

Comment (Alex009) [reply units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt]: just move from here `withReloadHandler: UICollectionViewReloadHandler = { collectionView, _, _ -> collectionView.reloadData() }`

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitTableViewDataSource.kt [hunk 19]
```diff
@@ -13,18 +13,26 @@ import platform.UIKit.row
...
 
+public typealias UITableViewReloadHandler = (UITableView, oldData: List<TableUnitItem>?, newData: List<TableUnitItem>?) -> Unit
+
 @ExportObjCClass
-class UnitTableViewDataSource(
-    private val tableView: UITableView
-) : NSObject(), UITableViewDataSourceProtocol {
+class UnitTableViewDataSource internal constructor(
+    private val tableView: UITableView,
+    private val reloadDataHandler: UITableViewReloadHandler
```
Comment (Alex009) [inline units/src/iosMain/kotlin/dev/icerock/moko/units/UnitTableViewDataSource.kt]: here default needed too

Comment (DevTchernov) [reply units/src/iosMain/kotlin/dev/icerock/moko/units/UnitTableViewDataSource.kt]: done

File: units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt (hunk 18 above)
Comment (DevTchernov) [reply units/src/iosMain/kotlin/dev/icerock/moko/units/UnitCollectionViewDataSource.kt]: done

File: units/src/iosMain/swift/Core/Units/UICellUnit.swift (hunk 10 above)
Comment (DevTchernov) [reply units/src/iosMain/swift/Core/Units/UICellUnit.swift]: fixed

**Pull request merged**